Feature/15 change query syntax

WojciechMazur opened the pull request:
Resolves #15 

Changed query syntax according to discussion at issue, e.q

```
Dao[T].all[.filter(...)*][.sortBy(...)][.reverse][.take(...)]()
```

1. Defined alias `query`  for `all`. Both of them creates new query builder. 
2. Query is materialized when using `apply` / `()` or defined alias `run`
1. Changed method of passing `implicit service: Service`. Current implementation did not allow to pass custom `Service` instance, e.q in tests. Also defined `default` and `noop` (mock) `Service` instances inside it's companion object
3. Query now returns `Stream[T]` instead of `Iterator[T]`. Such change allows to use new syntax as follows, where each case produces the same results

- `Dao[Type].all.filter(x).sortBy(y).reverse.sortBy(z).run()`
- `Dao[Type].all.filter(x).sortBy(y).reverse.run().sortBy(z)`
- `Dao[Type].all.filter(x).sortBy(y).run().reverse.sortBy(z)`
- `Dao[Type].all.filter(x).run().sortBy(y).reverse.sortBy(z)`
- `Dao[Type].all.run().filter(x).sortBy(y).reverse.sortBy(z)`


**Not applies changes:** 
1. Not renamed `Dao` to `Query`, as it allows for other operations, that is`unapply` allows fetching entity by Id and as `Dao` contains keyFactory

WojciechMazur committed:
Changed syntax according to issue #15. Allowed to pass different Service other then default

WojciechMazur committed:
Updated dependency layers:
Magnolia set to 1.2
Adversaria set to 1.3

propensive commented on src/core/mutatus.scala:
```diff
@@ -27,17 +27,12 @@ import scala.util.{Success, Try}
     /** saves the all case class as a Datastore entity in batch mode */
     def saveAll()(
-        implicit svc: Service,
+        implicit svc: Service = Service.default,
```
Would it be possible to move this default to an implicit defined withing the `Service` companion object? I think that defaults on implicit parameters have been known to be problematic, and defining it in the implicit scope of the companion object should have the same effect.

WojciechMazur replied:
I've changed it in first place due to ambiguous definitions, when I've tried to pass another implicit via method parameters. I've forget that I could just move it to companion object

propensive commented on src/core/mutatus.scala:
```diff
@@ -157,6 +152,26 @@ case class Service(readWrite: Datastore) {
+object Service {
+
+  /** provides a default instance of the GCP Datastore service */
+  lazy val default: Service = Service(
```
Make this one implicit.

propensive commented on src/core/mutatus.scala:
```diff
@@ -157,6 +152,26 @@ case class Service(readWrite: Datastore) {
+  )
+
+  /** Service instance which does not connect to any GCP Datastore, designed to be used in tests */
+  lazy val noop = Service {
```
If you move this one to a different object (or nest it inside another object) and make it implicit, you could get the "no-op" behaviour just with,
```
import Service.noop._
```

propensive commented on src/core/mutatus.scala:
```diff
@@ -257,29 +272,39 @@ case class LongId(id: Long) extends IdKey {
-  def query(implicit decoder: Decoder[T]): QueryBuilder[T] =
-    QueryBuilder[T](kind)
+  /** returns query builder with empty criteria which fetches all the values of this type stored in the GCP Platform */
+  def all: QueryBuilder[T] = QueryBuilder[T](kind)
```
I think we should choose either `all` or `query`, rather than duplicating it. The reason is that someone reading the code may see some code which refers to `all` and other code which refers to `query`. They would wonder what the difference is, when actually there is none. My preference would be for `all`.

WojciechMazur replied:
Should same apply to `apply` and it's alias `run`? In this case `apply` may be enough.

propensive replied:
Yes, please. Using `apply()` will be more concise, though it won't work if any of the query methods take implicit parameters. I don't think they do, but `run()` is also very short, and it's a very explicit indication to the user about what is in happening in Datastore and what is happening on-heap, so let's use `run()` and drop `apply()`.

propensive commented on src/core/mutatus.scala:
```diff
@@ -257,29 +272,39 @@ case class LongId(id: Long) extends IdKey {
 
+/** companion object for data access objects */
+object Dao {
+  implicit def apply[T](
```
I see why you've done it, though I'm a little uneasy about an implicit method that's also an `apply` method (and which also takes its own implicits). I'm not sure it's wrong...

propensive replied:
I was independently wondering whether we could provide the `Dao` functionality through an extension method on companion objects. But I think this can go into a separate issue/PR.

WojciechMazur replied:
In this case I've only added param `decoder:Decoder[T]` and moved companion object closer to case class definition (Before it was at the end of file, which made it bit annoying). 
I don't think it is wrong, although I don't remember seeing such such construct. I don't know what type of problems it may cause, but in tests it worked out well.

propensive replied:
Ok, so it's something I wrote! I probably had this argument with myself at the time, and then forgot about it. :) No need to change anything for this PR.

propensive commented on src/tests/EndToEndSpec.scala:
```diff
@@ -109,12 +109,12 @@ case class EndToEndSpec()(implicit runner: Runner) {
-  test("fetch entities - complex - long id")(
-    Dao[TestComplexLongId].all().toVector.sortBy(_.id)
-  ).assert(_ == longIdComplexEntities.sortBy(_.id))
+  // test("fetch entities - simple")(
+  //   Dao[TestSimpleEntity].all().toVector.sortBy(_.id)
+  // ).assert(_ == (simpleEntities ++ batchedSimpleEntities).sortBy(_.id))
+  // test("fetch entities - complex - long id")(
+  //   Dao[TestComplexLongId].all().toVector.sortBy(_.id)
+  // ).assert(_ == longIdComplexEntities.sortBy(_.id))
```
Is this work in progress?

WojciechMazur replied:
Ohh, I've must commented it out while catching a bug. Fixed

propensive commented on src/tests/EndToEndSpec.scala:
```diff
@@ -143,13 +143,14 @@ case class EndToEndSpec()(implicit runner: Runner) {
+      .filter(_.innerOpt.exists(_.int <= 8))
+      .sortBy(_.innerOpt.map(_.int))
+      .reverse
+      .drop(1)
+      .take(2)
```
Do you think we could also provide a `slice` method that combines `drop` and `take` in a single method?

WojciechMazur replied:
Sure, it shouldn't be a problem.

WojciechMazur replied:
Also I was thinking about supporting cursors, which are way more sufficient. When I was working in my previous project I was adjusting our API to use Elasticsearch API cursors pararell to offset based pagination inside GraphQL. 
In such case new method `after(<cursor>)` may be introduced.

propensive replied:
That would be useful, I think. Is there generally-accepted behaviour for how cursors work when the data is changing? For example, say you view the first 10 results from a query, and get a cursor to the last item shown. Then, elsewhere, the first five items are deleted, before you request the next ten records after the cursor.

WojciechMazur replied:
In case of offset pagination next request will result with original items 15-25 (since we would request `drop(10).take(10)` while items 1-5 have been deleted)
In case of cursor based implementation we would correctly get results 10-20 since we'll getting next 10 from last cursor which is not relative. In case if entity itself with last cursor would be deleted Datastore can still get correct next results.

If in same example if instead of deletion of elements 1-5 N items would be created in case of offset pagination N items from first result would appear in second result. In case of cursors such behavior should not happen, and result will be equal with the one from first example. 
https://cloud.google.com/datastore/docs/concepts/queries#cursors_and_data_updates

propensive commented on src/tests/QueryBuilderSpec.scala:
```diff
@@ -137,31 +131,33 @@ case class QueryBuilderSpec()(implicit runner: Runner) {
       .desc("innerClassOpt.deeperOpt.some2"),
     builder
-      .orderBy(_.desc(_.innerClassOpt.flatMap(_.deeperOpt).map(_.some2))) -> OrderBy
+      .sortBy(_.innerClassOpt.flatMap(_.deeperOpt).map(_.some2))
+      .reverse -> OrderBy
```
Wow... this is getting complex! ;)

propensive reviewed (changes requested):
Looks good. Just a couple of minor points, and questions. Anything which isn't easy to change should go into a new issue.

WojciechMazur committed:
After review fixes:
Removed aliases for Dao.all and QueryBuilder.run()
Removed default Service implicit parameters
Moved implicit default Service instance to it's companion object
Moved noopService to dedicated object Service.noop
Test fixes according to changes

WojciechMazur commented:
I'm wondering why CI fails at build. Can you take a look on it?
Used layerRef is `Qmdyx5CQ1EvxagyGMyceJoEv2MPR2c4oZnwEFaJioT1pBv`
```
Downloading Fury 0.12.3...done
Extracting Fury 0.12.3 into /home/runner/.local/share/fury/usr/0.12.3...done
Launching Fury 0.12.3 in standalone mode...
147.673 An IPFS operation timed out.
##[error]Process completed with exit code 1.
```

propensive commented on src/tests/EndToEndSpec.scala:
```diff
@@ -109,12 +109,12 @@ case class EndToEndSpec()(implicit runner: Runner) {
-  // test("fetch entities - complex - long id")(
-  //   Dao[TestComplexLongId].all().toVector.sortBy(_.id)
-  // ).assert(_ == longIdComplexEntities.sortBy(_.id))
+  test("fetch entities - simple")(
```
I would just make one tiny change to include `()` after every call to `run`, just to indicate to the reader that it's a side-effecting method.

propensive commented:
This is good to merge. If you get a chance to address my small comment about `run()` before you do, that would be great. :)

WojciechMazur committed:
Removed noop service as it would should be used outside QueryBuilderSpec. Added () to run to indicate side-effects
